stream_alert/rule_processor/classifier.py

```diff
@@ -224,54 +211,82 @@ def _parse(self, payload, data):
-            if parser.payload_type:
-                payload.type = parser.payload_type
+            # Get a list of parsed records
+            parsed_data = parser.parse(data)
 
+            LOGGER.debug('schema: %s', schema)
             if parsed_data:
-                logger.debug('log name: %s', log_name)
-                logger.debug('parsed_data: %s', parsed_data)
-                typed_data = []
-                for data in parsed_data:
-                    # Convert data types per the schema
-                    # Use the parser.schema due to updates caused by
-                    #   configuration settings such as envelope and optional_keys
-                    typed_data.append(self._convert_type(data, parser.schema, options))
-
-                if typed_data:
-                    payload.log_source = log_name
-                    payload.type = parser_name
-                    payload.records = typed_data
-                    return True
-        return False
+                passes.append((log_name, parser, parsed_data))
+
+        if not passes:
+            return False
 
-    def _convert_type(self, parsed_data, schema, options):
+        valid_parse = passes[0]
+        if len(passes) > 1:
+            matched_parses = []
+            for i, valid_parses in enumerate(passes):
+                parser = valid_parses[1]
+                for data in valid_parses[2]:
+                    if parser.matched_log_pattern(data, parser.options.get('log_patterns', {})):
+                        matched_parses.append(passes[i])
+                        break
+                    else:
+                        LOGGER.debug('log pattern matching failed for schema: %s', parser.schema)
+
+            if len(matched_parses) != 0:
+                valid_parse = matched_parses[0]
+                if len(matched_parses) > 1:
+                    LOGGER.error('log patterns matched for multiple schemas: %s',
+                                 ', '.join(name for name, _, _ in matched_parses))
+                    LOGGER.error('proceeding with schema for: %s', valid_parse[0])
+            else:
+                LOGGER.error('log classification matched for multiple schemas: %s',
+                             ', '.join(name for name, _, _ in passes))
+                LOGGER.error('proceeding with schema for: %s', passes[0][0])
+
+        log_name, parser, parsed_data = valid_parse[0], valid_parse[1], valid_parse[2]
+        LOGGER.debug('log_name: %s', log_name)
+        LOGGER.debug('parsed_data: %s', parsed_data)
+
+        typed_data = []
+        for data in parsed_data:
+            # Convert data types per the schema
+            # Use the parser.schema due to updates caused by
+            #   configuration settings such as envelope and optional_keys
+            converted_data = self._convert_type(data, parser.type(), parser.schema, parser.options)
+            if not converted_data:
+                payload.valid = False
```

Comment:
No - `payload.valid_source` was an unnecessary var. `payload.valid` is different and has been there to store if any log in a payload does not get classified properly.